federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+
+		// TODO: We could do a dry-run here if wanted to avoid the spurious cloud-calls & events when we restart
+
+		// The DNS records don't exist yet, so create them.
+		// TODO: quinton: Split this up and handle each record separately.
+		//                Beware though, creating distant ones before near ones can lead to race condidtions on name resolution.
+		//                (e.g. client local resolution fails as DNS record doesn't exist yet, but then tries remote resolution and that succeeds,
+		//                so it gets stuck for a while on a remote service shard.
+		//                Solutions: either use transactions (where providers suport them, e.g. Google Cloud DNS) or just make
+		//                sure that the creation sequence is zone-local, region-local, sub-continent local etc.  The latter probably
+		//                works for all cases, so transactions are probably not necessary.
+		//                Actually, on further thought, the race condition exists anyway - neither of the above tow really solve it.
+		//                Ho hum, have to think harder.
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatingDNSRecords", "Creating DNS Records")
+		err := s.ensureDNSRecords(clusterName, service)
+		if err != nil {
+			return fmt.Errorf("Failed to create DNS records for service %s: %v", key, err), retryable
+		}
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatedDNSRecords", "Created DNS records")
+	}
+
+	serviceInMap, ok := s.cache.get(service.Name)
+	if !ok {
+		return fmt.Errorf("service cache lost")
+	}
+	// TODO: remove dup ip info?
+	serviceInMap.appliedState.Spec.ExternalIPs = append(serviceInMap.appliedState.Spec.ExternalIPs, service.Spec.ExternalIPs)
+	// write back to federation apiserver
+
+	// TODO to check LB info equals
+	//if !api.LoadBalancerStatusEqual(previousState, &service.Status.LoadBalancer) { // TODO: quinton: only interested in the DNS record here.
+	//	// Not even sure it needs updating - check.
+	//} else {
+	_, err := s.federationClient.Core().Services().Update(serviceInMap)
+	if err != nil {
+		return fmt.Errorf("fail to update federation service")
+	}
+	//	glog.V(2).Infof("Not persisting unchanged LoadBalancerStatus to registry.") // TODO: quinton: check/fix this.
+	//}
+
+	return nil, notRetryable
+}
+
+// Returns whatever error occurred along with a boolean indicator of whether it
+// should be retried.
+func (s *ServiceController) ensureDNSRecordsExist(key string, fedService, appliedState *v1.Service) (error, bool) {
+	// Clone federation service, and create them in underlying k8s cluster
+	clone, err := conversion.NewCloner().DeepCopy(fedService)
+	if err != nil {
+		return nil, err
+	}
+	service, ok := clone.(*v1.Service)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected service cast error : %v\n", service)
+	}
+	// to generate service name, we need a api.ObjectMeta instead of v1
+	meta := &api.ObjectMeta{}
+	meta.GenerateName = service.ObjectMeta.Name + "-"
```

Comment:
Yes, apologies for the change of plan Jesse.  I would suggest that we keep
labelling the underlying objects (services in this case) to retain
consistency with other object types (e.g. deployment). It also allows
convenient operations in future like 'list all services labelled x'. We
can't use get for that, as get can only return one object.  Hope that makes
sense?
On May 13, 2016 12:23 AM, "Nikhil Jindal" <notifications@github.com> wrote:

> In federation/pkg/controller/service/servicecontroller.go
> <https://github.com/kubernetes/kubernetes/pull/25517#discussion_r63142020>
> :
>
> > +
> > +// Returns whatever error occurred along with a boolean indicator of whether it
> > +// should be retried.
> > +func (s *ServiceController) ensureDNSRecordsExist(key string, fedService, appliedState *v1.Service) (error, bool) {
> > +	// Clone federation service, and create them in underlying k8s cluster
> > +	clone, err := conversion.NewCloner().DeepCopy(fedService)
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	service, ok := clone.(*v1.Service)
> > +	if !ok {
> > +		return nil, fmt.Errorf("Unexpected service cast error : %v\n", service)
> > +	}
> > +	// to generate service name, we need a api.ObjectMeta instead of v1
> > +	meta := &api.ObjectMeta{}
> > +	meta.GenerateName = service.ObjectMeta.Name + "-"
>
> Yes. @quinton-hoole <https://github.com/quinton-hoole> and I had another
> discussion this morning and he convinced me that using the same name in
> federation and k8s services is better that using the spec hash label. Its
> easier for users to understand. Re: how to handle the situation where there
> might be an existing k8s service with the same name as a newly created
> federation service, we decided that the federation service can adopt that
> existing k8s service. If required, federation service controller will
> update the spec of that existing k8s service to match the spec of
> federation service.
> Ref #23848 (comment)
> <https://github.com/kubernetes/kubernetes/issues/23848#issuecomment-218867282>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25517/files/7401226d3cb8451664672c4092084e1b4f6f4f3f#r63142020>
>